framework/include/materials/Material.h

```diff
@@ -258,6 +259,25 @@ class Material :
    */
   virtual QpData * createData();
 
+  /**
+   * Get a apecial MaterialUserObject which can be explicitly called by the material
+   * during the material property calculation.
+   * @param name The name of the parameter key of the user object to retrieve
+   * @return The user object with name associated with the parameter 'name'
+   */
+  template<class T>
+  const T & getMaterialUserObject(const std::string & name);
```

Comment:
I like this idea. It'll avoid exposing `tid` to the user so people won't be able to screw up the retrieval of normal UserObjects.